### PR TITLE
feat: [CDS-70210]: Added ManifestConfigurations object in service yaml.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Updated service yaml to

```
serviceDefinition:
    type: Kubernetes
    spec:
      manifestConfigurations:
        primaryManifestRef: <+input>
      manifests:
        - manifest:
```

Screenshot 2023-06-23 at 10 37 33 AM
Screenshot 2023-06-23 at 10 37 43 AM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/49644)

### CODE REVIEW

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger ss

**Comment by GoelPratyush99** [comment]
trigger ut3, yt9

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/ManifestConfigurations.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import org.springframework.data.annotation.TypeAlias;
+
+@OwnedBy(HarnessTeam.CDP)
+@Data
+@NoArgsConstructor
+@SimpleVisitorHelper(helperClass = ManifestConfigurationsVisitorHelper.class)
+@TypeAlias("ManifestConfigurations")
```
Comment (acasian-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/ManifestConfigurations.java:33]: i think it was asked already, is it `manifestConfigurations` or `manifestConfiguration`?

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java [hunk 2]
```diff
@@ -51,6 +52,7 @@ public class KubernetesServiceSpec implements ServiceSpec, Visitable {
   List<ManifestConfigWrapper> manifests;
   List<ConfigFileWrapper> configFiles;
   List<ServiceHookWrapper> hooks;
+  ManifestConfigurations manifestConfigurations;
```
Comment (acasian-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:55]: do we have any UTs covering the specs?

**Comment by GoelPratyush99** [comment]
trigger uts

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java (hunk 2 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:55]: No we do not have it for specs

**Comment by GoelPratyush99** [comment]
trigger uts

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/ManifestConfigurations.java (hunk 1 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/manifestConfigs/ManifestConfigurations.java:33]: Yes got confirmation for rohan. We are going ahead with ManifestConfigurations

**Comment by GoelPratyush99** [comment]
trigger review

**Comment by GoelPratyush99** [comment]
trigger uts

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java (hunk 2 above)
Comment (acasian-harness) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:55]: @abosii-harness shall we consider UT's for specs also?

**Comment by TarunUba1** [comment]
For safety purpose, did you test the normal flow? and also were you able to get these variables in stepElementParameters?

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java [hunk 3]
```diff
@@ -80,6 +82,8 @@ public VisitableChildren getChildrenToWalk() {
       hooks.forEach(hook -> children.add("hooks", hook));
     }
 
+    children.add("manifestConfigurations", manifestConfigurations);
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:85]: Should we add null check?

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java (hunk 2 above)
Comment (abosii-harness) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:55]: I guess there is no point for this

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/NativeHelmServiceSpec.java [hunk 4]
```diff
@@ -74,6 +76,8 @@ public VisitableChildren getChildrenToWalk() {
       hooks.forEach(hook -> children.add("hooks", hook));
     }
 
+    children.add("manifestConfigurations", manifestConfigurations);
```
Comment (abosii-harness) [inline 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/NativeHelmServiceSpec.java:79]: Should we add null check?

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java (hunk 3 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/KubernetesServiceSpec.java:85]: I think its fine. We do not add it to artifacts field (non - mandatory)

File: 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/NativeHelmServiceSpec.java (hunk 4 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/cdng/service/beans/NativeHelmServiceSpec.java:79]: I think its fine. We do not add it to artifacts field (non - mandatory)

**Review by GoelPratyush99 — COMMENTED** [review]
Yes, I had test it on normal flow it worked fine. In stepElementParameters we will not see this field as manifestConfiguration is not manifest spec.

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49644#-:-NYs1nQg9tXV1pbGi5So:b-trr9kp)* status: 7 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 2 approvals obtained.

**Review by abosii-harness — APPROVED** [review]

**Review by TarunUba1 — DISMISSED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abosii-harness — APPROVED** [review]

**Review by TarunUba1 — APPROVED** [review]

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 5 of 5 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/49644)* status: 7 file(s) to review. 4. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
trigger review

**Comment by GoelPratyush99** [comment]
trigger review

**Pull request merged**